### DISCUSSION TITLE
Support Distribute with TestFlight Internal Only

### DISCUSSION THREAD

**Discussion by effx13** [fastlane/fastlane#21525] [Ideas]
Hello

In the recently released Xcode 15, there is a new method to upload to TestFlight Internal Only.
![image](https://github.com/fastlane/fastlane/assets/29156835/6fe6271e-dce7-4e39-a3c5-56bc2b524d6c)

This method can internal testing only. It can't be submitted for external testing or the App Store, but it is processed on Apple servers more quickly than common builds, so it make internal testing faster.

Currently, `upload_to_testflight` doesn't seem to have this option. 
Does anybody out there have this working, or have any ideas?

Thanks!

Related documents:
- https://developer.apple.com/documentation/xcode/distributing-your-app-for-beta-testing-and-releases
- https://developer.apple.com/help/app-store-connect/test-a-beta-version/view-build-status-and-metrics
- https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes

**Comment by markus-mohemian** [comment]
You can already achieve that by setting `export_options` in gym:
e.g.
```
gym(
    ...
    export_options: {
        testFlightInternalTestingOnly: true
    }
)
```
Keep in mind you need Xcode 15 for that.

**Reply by stanbohonos** [reply]
This setting doesn't work over fastlane. 

`[altool] 2024-02-17 02:08:40.969 *** Error: Asset validation failed Invalid Provisioning Profile. The provisioning profile included in the com.xxxxx bundle  is invalid. [Missing code-signing certificate]. A distribution provisioning profile should be used when uploading apps to App Store Connect. (ID: 284eb807-aa8d-4fe7-8b4f-6ec26c0bfbe9) (90161)`

But for such distribution Internal only, as I understand it, developer profiles must be used. 

At least that's how I publish from xcode (developer profiles and cert).

**Comment by chanphiromsok** [comment]
can you share me which package are you using or share me the configuration ?

**Reply by zakharchenkoAndrii** [reply]
If you will run `xcodebuild --help` command you will find 

```
Available keys for -exportOptionsPlist:

        testFlightInternalTestingOnly : Bool

                When enabled, this build cannot be distributed via external TestFlight or the App Store. This is recommended for pull requests, development branches, and other builds that are not suitable for external distribution.
```
So the thing you have to do is to add 
`testFlightInternalTestingOnly` key/value to your .plist

```
    <key>testFlightInternalTestingOnly</key>
    <true/>
```

**Comment by Fuller43** [comment]
Thank you for being so generous with your knowledge. [Liteblue](https://liteblue.cloud) It's truly appreciated.

**Comment by wei18** [comment]
Internal builds are not being tagged as 'internal' on TestFlight
ref: https://forums.developer.apple.com/forums/thread/759317
